File: Python/Arrays and Hashing/ProductArrayExceptSelf.py
# 238. Product of Array Except Self
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefixs = []
        sufix = []

        # Calculate all the prefix
        for i in range(len(nums)):
            if i == 0:
                prefixs.append(1)
                continue
            aux = 1
            for j in range(i):
                aux = aux * nums[j]
            prefixs.append(aux)

        # Calculate all the sufix
        for i in range(len(nums)):
            if i == len(nums) - 1:  # last element
                sufix.append(1)
                continue
            aux = 1
            j = len(nums)
            while j > 0:
                aux = aux * nums[j]
                j = j - 1

        return nums

File: cpp/Arrays and Hashing/ProductArrayExceptSelf.cpp
#include <vector>
#include <iostream>
#include <set>
#include <unordered_set>
using namespace std;

class Solution {
    public:
        vector<int> productExceptSelf(vector<int>& nums) {
            vector<int> res;    
            int len = nums.size();        

            // calculate the list of prefix
            vector<int> prefix;
            int aux_prefix = 1;
            for (int i = 0; i < len; i++){
                prefix.push_back(aux_prefix);
                aux_prefix *= nums[i];
            }
            // calculate the list of sufix
            vector<int> sufix;
            int aux_sufix = 1;
            for(int i = len - 1; i > 0 ; i--){
                sufix.push_back(aux_sufix);
                aux_sufix *= nums[i];
            }

            for(int i = 0; i < len; i++){
                res.push_back(prefix[i] * prefix[i]);
            }
            return res;
        }
};